src/solvers/refinement/string_refinement.h

```diff
@@ -164,39 +160,57 @@ class string_refinementt: public bv_refinementt
 /// the left to initialize the unconcretized indices of concrete_array.
 /// \param concrete_array: the vector to populate
-/// \param initialized: the vector containing the indices of the concretized
-///   values
-/// \param last_concretized: initial value of the last concretized index
-template <typename T1, typename T2>
-void string_refinementt::pad_vector(
-  std::vector<T1> &concrete_array,
-  std::set<T2> &initialized,
-  T1 last_concretized) const
+/// \param initialized: set containing the indices of already concrete values
+template <typename T>
+void fill_in_vector(
+  std::vector<T> &concrete_array,
+  const std::set<std::size_t> &initialized)
+{
+  PRECONDITION(!initialized.empty());
+  auto it=initialized.begin();
+  PRECONDITION(*initialized.rbegin()==concrete_array.size()-1);
+
+  for(std::size_t j=0; j<concrete_array.size(); j++)
+    if(*it==j)
+      ++it;
+    else
+      concrete_array[j]=concrete_array[*it];
+}
+
+/// Utility function for concretization of strings. Copies concretized values to
+/// the left to initialize the unconcretized indices of concrete_array.
+/// \param initial_map: map containing the values of already concrete values
+/// \return a populated vector with value from initialized
+template <typename T>
+std::vector<T> fill_in_map_as_vector(
+  const std::map<std::size_t, T> &initial_map)
 {
+  if(initial_map.empty())
+    return std::vector<T>();
+
+  std::size_t last_index=initial_map.rbegin()->first;
+  std::vector<T> result(last_index+1);
   // Pad the concretized values to the left to assign the uninitialized
-  // values of result. The indices greater than concretize_limit are
-  // already assigned to last_concretized.
-  for(auto j=initialized.rbegin(); j!=initialized.rend();)
+  // values of result.
+  for(auto pair=initial_map.rbegin(); pair!=initial_map.rend();)
   {
-    size_t i=*j;
-    // The leftmost index to pad is the value + 1 of the next element in
-    // 'initialized'. Since we cannot use the binary '+' operator on set
-    // iterators, we must increment the iterator here instead of in the
-    // for loop.
-    j++;
-    size_t leftmost_index_to_pad=(j!=initialized.rend()?*(j)+1:0);
-    // pad until we reach the next initialized index (right to left)
-    while(i>leftmost_index_to_pad)
-      concrete_array[(i--)-1]=last_concretized;
-    INVARIANT(
-      i==leftmost_index_to_pad,
-      string_refinement_invariantt("Loop decrements i until it is not greater "
-        " than leftmost_index_to_pad"));
-    if(i>0)
-      last_concretized=concrete_array[i-1];
+    const std::size_t i=pair->first;
+    const T value=pair->second;
+    // We must increment the iterator here instead of in the for loop so that
+    // we can get the leftmost index to pad
+    pair++;
+    const std::size_t leftmost_index_to_pad=
+      pair!=initial_map.rend()?pair->first+1:0;
+    // pad down to the leftmost index to pad
+    for(std::size_t j=i; j+1>leftmost_index_to_pad; j--)
```

Comment:
Comment something like `Conditon would be j>=leftmost_index_to_pad, but it's an unsigned type and leftmost_index_to_pad might be 0.`